### PR TITLE
Add an optional callback parameter to get notified of intercepted messages

### PR DESCRIPTION
**Pull request by janstrohbeck**
This allows the data source to get notified when a message was received at the orchestrator. I use this to save all received messages to JSON files, to compare the outputs of nodes across runs. Could also potentially be used for other purposes. Messages may have to be decoded in case of raw subscriptions, which is why the topic type is also passed to the callback.

### CODE REVIEW

File: ros2/orchestrator/orchestrator/orchestrator_lib/orchestrator.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ class Orchestrator:
 
     def __init__(self, ros_node: RosNode, executor: Executor, node_config: List[NodeModel],
                  logger: Optional[RcutilsLogger] = None, timing_analysis: bool = False,
-                 state_sequence_recording: bool = False) -> None:
+                 state_sequence_recording: bool = False, intercepted_topic_callback: Callable[[str, type, any], None] = None) -> None:
```
Comment (ottojo) [inline ros2/orchestrator/orchestrator/orchestrator_lib/orchestrator.py]:
```suggestion
                 state_sequence_recording: bool = False, intercepted_topic_callback: Optional[Callable[[str, type, Any], None]] = None) -> None:
```

File: ros2/orchestrator/orchestrator/orchestrator_lib/orchestrator.py [hunk 2]
```diff
@@ -288,6 +295,7 @@ def dataprovider_publish(self, topic: TopicName, message: Any) -> None:
         if topic not in self.interception_subs.keys():
             self.l.info(f"Got a message on topic {topic} but we are not subscribed to this input. Ignoring.")
             return
+        self.topic_types[topic] = type(message)
```
Comment (ottojo) [inline ros2/orchestrator/orchestrator/orchestrator_lib/orchestrator.py]: why is this necessary here?

**Review by ottojo — CHANGES_REQUESTED** [review]
please rebase or merge develop to run CI on the PR

File: ros2/orchestrator/orchestrator/orchestrator_lib/orchestrator.py (hunk 2 above)
Comment (janstrohbeck) [reply ros2/orchestrator/orchestrator/orchestrator_lib/orchestrator.py]: You're right, it isn't. I removed it.

**Review by ottojo — APPROVED** [review]
